client/src/test/java/io/pravega/client/tables/impl/KeyValueTableTestBase.java

```diff
@@ -0,0 +1,597 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.tables.impl;
+
+import io.pravega.client.stream.Serializer;
+import io.pravega.client.stream.impl.UTF8StringSerializer;
+import io.pravega.client.tables.BadKeyVersionException;
+import io.pravega.client.tables.IteratorItem;
+import io.pravega.client.tables.IteratorState;
+import io.pravega.client.tables.KeyValueTable;
+import io.pravega.client.tables.TableEntry;
+import io.pravega.client.tables.TableKey;
+import io.pravega.client.tables.Version;
+import io.pravega.common.util.AsyncIterator;
+import io.pravega.test.common.AssertExtensions;
+import io.pravega.test.common.LeakDetectorTestSuite;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+import java.util.function.BiPredicate;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import lombok.AccessLevel;
+import lombok.Cleanup;
+import lombok.Getter;
+import lombok.val;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * Base test suite for anything testing {@link KeyValueTable}s. This covers core functionality for {@link KeyValueTable}s
+ * and currently applies both to {@link KeyValueTableImplTests} (using mocked Controller and Segment Store) and
+ * `io.pravega.test.integration.KeyValueTableImplTests` (using real Segment Store and Wire Protocol).
+ */
+public abstract class KeyValueTableTestBase extends LeakDetectorTestSuite {
+    //region Members
+
+    protected static final String NULL_KEY_FAMILY = "[NULL]"; // Used for HashMap keys.
+    protected static final Serializer<Integer> KEY_SERIALIZER = new IntegerSerializer();
+    protected static final Serializer<String> VALUE_SERIALIZER = new UTF8StringSerializer();
+    private static final int DEFAULT_SEGMENT_COUNT = 4;
+    private static final int DEFAULT_KEY_FAMILY_COUNT = 100;
+    private static final int DEFAULT_KEYS_PER_KEY_FAMILY = 10;
+    @Getter(AccessLevel.PROTECTED)
+    private List<String> keyFamilies;
+
+    //endregion
+
+    // Setup and configuration
+
+    @Before
+    public void setup() throws Exception {
+        int count = getKeyFamilyCount();
+        this.keyFamilies = new ArrayList<>();
+        this.keyFamilies.add(null); // No key family.
+        for (int i = 0; i < count; i++) {
+            this.keyFamilies.add(String.format("KF[%d]", i));
+        }
+
+        this.keyFamilies = Collections.unmodifiableList(this.keyFamilies);
+    }
+
+    @Override
+    protected int getThreadPoolSize() {
+        return 3;
```

Comment:
Just to confirm, the tests pass if this is 1 right?

Reply:
Yes they do. But the tests run somewhat faster with 3 given that at least in one implementation of this class we run the entire segment store.